File: ruby/testcode/ldapi_047.rb
require_relative 'shared'

begin; require 'net/ldap'; rescue LoadError; end
ldap = Net::LDAP.new(host: 'ldap.corp.com', port: 389) rescue nil

LDAP_SCOPE = Net::LDAP::SearchScope_WholeSubtree rescue 2

# vuln-code-snippet start ruby_ldapi_scope_constant
def search_whole_subtree(req)
  filter = Net::LDAP::Filter.eq('uid', req.param('uid'))
  ldap.search(base: 'dc=corp,dc=com', scope: LDAP_SCOPE, filter: filter) # vuln-code-snippet safe-line ruby_ldapi_scope_constant
  BenchmarkResponse.json({ ok: true })
end
# vuln-code-snippet end ruby_ldapi_scope_constant

File: rust/testcode/ldapi_047.rs
//! CWE-90: User input used only in response display string; LDAP filter is hardcoded and never tainted.

fn ldap_search(base: &str, filter: &str) -> String {
    // In production: ldap3::LdapConn::new("ldap://...").search(base, Scope::Subtree, filter, vec!["cn", "mail"])
    format!("LDAP search in {} with filter {}", base, filter)
}

// vuln-code-snippet start testcodeLdapi047
pub fn handle(req: &super::shared::BenchmarkRequest) -> super::shared::BenchmarkResponse {
    let display_name = req.param("display_name");
    let filter = "(&(objectClass=person)(active=true))";
    let ldap_result = ldap_search("dc=example,dc=com", filter); // vuln-code-snippet target-line testcodeLdapi047
    let response = format!("Hello {}, results: {}", display_name, ldap_result);
    super::shared::BenchmarkResponse::ok(&response)
}
// vuln-code-snippet end testcodeLdapi047
